Spectrum instrumentation ADC as fast_counter hardware

takuya-ulm opened the pull request:
<!--- Provide a general short and descriptive title above -->
## Description
<!--- Describe your changes in detail -->
AD converter from spectrum instrumentation is implemented as fast_counter hardware.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
AD card from spectrum instrumentation is newly implemented.
The class structure is organized for robustness.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
This hardware is tested in the pulsed toolchain.
This change does not affect the rest of the code.

## Screenshots (only if appropriate, delete if not):

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Put an 'x' in all the boxes that apply, e.g. "[x]". -->
<!--- You can also create this PR and afterwards check the boxes by clicking on them. -->
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change (Causes existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an 'x' in all the boxes that apply, e.g. "[x]". -->
<!--- If you're unsure about any of these, ask. -->
<!--- You can also create this PR and afterwards check the boxes by clicking on them. -->
- [x] My code follows the code style of this project.
- [ ] I have documented my changes in `/docs/changelog.md`.
- [ ] My change requires additional/updated documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added/updated the config example for any module docstrings as necessary.
- [ ] I have checked that the change does not contain obvious errors
(syntax, indentation, mutable default values, etc.).
- [ ] I have tested my changes using 'Load all modules' on the default dummy configuration.
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

TobiasSpohn commented on src/qudi/gui/time_series/time_series_gui.py:
For a new hardware implementation this file should not be changed.

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_commands/buffer_commands.py:
```diff
@@ -0,0 +1,127 @@
+You should have received a copy of the GNU Lesser General Public License along with qudi.
+If not, see <https://www.gnu.org/licenses/>.
+"""
+from pyspcm import *
```
Import only the function and classes that are needed in the file, not the whole package.

Neverhorst replied:
Since this package is heavily used, it would also be viable to import the package as a whole without "*" in order to avoid cluttering the module namespace, i.e. `import pyspcm`

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_commands/card_commands.py:
```diff
@@ -0,0 +1,75 @@
+You should have received a copy of the GNU Lesser General Public License along with qudi.
+If not, see <https://www.gnu.org/licenses/>.
+"""
+from pyspcm import *
```
Import only the function and classes that are needed in the file, not the whole package.

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_commands/configure_commands.py:
```diff
@@ -0,0 +1,549 @@
+"""
+import ctypes
+from functools import reduce
+from pyspcm import *
```
Import only the function and classes that are needed in the file, not the whole package.

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_commands/si_commands.py:
```diff
@@ -0,0 +1,128 @@
+        curr_ts_avail_reps = self.cmd.process.get_ts_avail_user_reps()
+        return min(curr_data_avail_reps, curr_ts_avail_reps)
+
+    def wait_avail_data(self):
+        initial_time = time.time()
+        curr_avail_reps = self.get_curr_avail_reps()
+        while curr_avail_reps == 0:
+            curr_avail_reps = self.get_curr_avail_reps()
+            current_time = time.time() - initial_time
+            if current_time > 10:
+                time_out = True
+                return time_out
+            time.sleep(self.wait_time_interval)
+        time_out = False
+        return time_out
```
This function is never used

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_main.py:
```diff
@@ -0,0 +1,411 @@
+import numpy as np
+
+from enum import IntEnum
+from pyspcm import *
+from spcm_tools import *
```
excplicitly import only the necessary classes and functions, not the whole module.

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_main.py:
```diff
@@ -0,0 +1,411 @@
+                3 = paused
+                -1 = error state
+        """
+        return self._internal_status
```
As far as I understood, the _internal_status is just a python variable, not associated with any response from the hardware itself. Thus during a measurement an error could occur with the card but when checking with get_status() one would still get the result that the card is running as _internal_status was set to this during the start of the measurement.
Maybe the interface allows for a status query of the card?

takuya-ulm replied:
It would be great if this error state is used when the error from the hardware is caught. I added this as TODO.

TobiasSpohn commented on src/qudi/hardware/fast_adc/spectrum/si_commands/si_commands.py:
```diff
@@ -0,0 +1,128 @@
+    def wait_avail_data(self):
+        initial_time = time.time()
+        curr_avail_reps = self.get_curr_avail_reps()
+        while curr_avail_reps == 0:
+            curr_avail_reps = self.get_curr_avail_reps()
+            current_time = time.time() - initial_time
+            if current_time > 10:
+                time_out = True
+                return time_out
+            time.sleep(self.wait_time_interval)
```
This might also lock the thread for a significant amount of time, making it unresponsive. If wait_time_interval is too long this might cause problems with stopping the acquisition.

takuya-ulm replied:
I removed this wait function from the measurement loop(si_commander).
Now, instead of waiting in another while loop(wait_new_trig_reps), it returns in the measurement loop.

``` 
def command_process(self):
        """
        Command the main process dependent on the repetitions of unprocessed data.
        """

        trig_reps = self.cmd.process.get_trig_reps()
        processed_reps = self.processor.data.avg.num_rep
        unprocessed_reps = trig_reps - processed_reps
        if self._reps != 0 and processed_reps >= self._reps:
            return
```

TobiasSpohn reviewed (commented):
Hi Takuya,

Great code. I added some small suggestions in the comments. 
In my comments about the while-loops, making the thread unresponsive: It might be okay to leave it as is, if the time spent in the loop is very short, so the thread is not blocked for too long.

Neverhorst commented:
Where does the `pyspcm` dependency come from? It is not listed in `setup.py`. 
Is this derivative work of [this project](https://github.com/vuthalab/spectrum-awg/tree/master)? If so, this needs proper attribution and copyright/license disclaimers.

takuya-ulm commented:
> Where does the `pyspcm` dependency come from? It is not listed in `setup.py`. Is this derivative work of [this project](https://github.com/vuthalab/spectrum-awg/tree/master)? If so, this needs proper attribution and copyright/license disclaimers.

Spectrum instrumentation originally distributed some python codes with an USB stick. Recently, they shared the codes on github with MIT license:
[https://github.com/SpectrumInstrumentation/spcm](url)
Now I added this package in setup.py and modified the imports.

takuya-ulm reviewed (commented):
@TobiasSpohn 

Thank you for your reviewing. I modified the codes and replied to your comments.
Can you have a look on them?

TobiasSpohn commented on src/qudi/gui/time_series/time_series_gui.py:
```diff
@@ -326,7 +326,7 @@ def update_channel_settings(self, enabled, averaged):
                 self._mw.trace_plot_widget.setLabel('right',
                                                     self._channels_per_axis[1][0],
-                                                    units=different_units[1])
+                                rev                    units=different_units[1])
```
Is this "rev" supposed to be here?

TobiasSpohn reviewed (commented):
Thanks for the modified code, it looks good from my side. I only have the one comment in time_series_gui.py, which I think might be a syntax error.

timoML commented on src/qudi/hardware/fast_adc/spectrum/si_main.py:
```diff
@@ -0,0 +1,418 @@
+
+        If the hardware does not support these features, the values should be None
+        """
+        with self._threadlock:
```
This thread lock locks sketchy. It is not taken when writing to the same data in DataProcess.
